source/opt/types.cpp

```diff
@@ -383,63 +384,72 @@ void SampledImage::GetExtraHashWords(
   image_type_->GetHashWords(words, seen);
 }
 
-Array::Array(Type* type, uint32_t length_id, uint32_t spec_id)
-    : Type(kArray),
-      element_type_(type),
-      length_id_(length_id),
-      length_spec_id_(spec_id),
-      length_constant_type_(nullptr),
-      length_constant_words_() {
+Array::Array(const Type* type, const Array::LengthInfo& length_info)
+    : Type(kArray), element_type_(type), length_info_(length_info) {
+  assert(type != nullptr);
   assert(!type->AsVoid());
-  assert(spec_id != 0u);
-}
-
-Array::Array(Type* type, uint32_t length_id, const Type* constant_type,
-             Operand::OperandData constant_words)
-    : Type(kArray),
-      element_type_(type),
-      length_id_(length_id),
-      length_spec_id_(0u),
-      length_constant_type_(constant_type),
-      length_constant_words_(constant_words) {
-  assert(!type->AsVoid());
-  assert(constant_type && constant_type->AsInteger());
+  assert(length_info.opcode == SpvOpSpecConstant ||
+         length_info.opcode == SpvOpSpecConstantOp ||
+         length_info.opcode == SpvOpConstant);
+  // A spec constant leaf value must have a spec id.
```

Comment:
This sounds like a strange requirement. Maybe clarify that the spec constant must have a spec id word even if it was not decorated.

Reply:
I assumed an OpSpecConstant is always decorated with a SpecId.  I realize that doesn't have to be the case.  I'll have to rethink the logic in Array::IsSameImpl (and related)

Reply:
Addressed in new commit

Reply:
Hey:  @pierremoreau 
This might partly disable the effect you were looking for in the linker.  If two spec constants have the same spec id but different literal values, then they should compare as unequal.

Reply:
cc: @s-perron who reviewed the original change.